aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SyncProcessor.java

```diff
@@ -166,28 +168,41 @@ private SyncTime filterOutOldSyncTimes(SyncTime lastSyncTime) throws DataStoreEx
      *     of time, then request a *delta* sync. If the last sync time is outside a recent window of time,
      *     perform a *base* sync. A base sync is preformed by passing null.
-     *  3. Update the
+     *  3. Continue fetching paged results until !hasNextPage() or we have synced the max records.
      * @param modelClass The model class to sync
-     * @param <T> The type of model to sync
+     * @param <T> The type of model to sync.
      * @return An {@link Single} which emits sync content, on success, {@link DataStoreException} on failure
      */
     private <T extends Model> Single<Iterable<ModelWithMetadata<T>>> syncModel(
             Class<T> modelClass, SyncTime syncTime) {
         final Long lastSyncTimeAsLong = syncTime.exists() ? syncTime.toLong() : null;
         return Single.<Iterable<ModelWithMetadata<T>>>create(emitter -> {
-            final Cancelable cancelable =
-                appSync.sync(modelClass, lastSyncTimeAsLong, metadataEmitter(emitter), emitter::onError);
-            emitter.setDisposable(AmplifyDisposables.fromCancelable(cancelable));
+            final Integer syncPageSize = dataStoreConfigurationProvider.getConfiguration().getSyncPageSize();
+            GraphQLRequest<PaginatedResult<ModelWithMetadata<T>>> request =
+                    appSync.buildSyncRequest(modelClass, lastSyncTimeAsLong, syncPageSize);
+            syncPage(request, emitter, new HashSet<>());
         }).doOnSuccess(results ->
             LOG.debug("Successfully sync'd down cloud state for model type = " + modelClass.getSimpleName())
         ).doOnError(failureToSync ->
             LOG.warn("Failed to sync down cloud state for model type = " + modelClass.getSimpleName(), failureToSync)
         );
     }
 
-    private static <T extends Model> Consumer<GraphQLResponse<Iterable<ModelWithMetadata<T>>>> metadataEmitter(
-        SingleEmitter<Iterable<ModelWithMetadata<T>>> singleEmitter) {
-        return resultFromEndpoint -> {
+    /**
+     * Recursively fetches each page for a sync, until there are no more pages available, or we have fetched the maximum
+     * configured number of records to sync (syncMaxRecords).
+     * @param request GraphQLRequest object for the sync, obtained from appsync.buildFirstPageSyncRequest, or from
+     *                response.getData().getRequestForNextResult() for subsequent requests.
+     * @param singleEmitter A SingleEmitter which emits an Iterable&lt;ModelWithMetadata&gt;, a concatenation of the
+     *                      results from all pages.
+     * @param emittedValue a Set&lt;ModelWithMetadata&lt;T&gt;&gt; containing the concatenated results of all requests.
+     * @param <T> The type of model to sync.
+     */
+    private <T extends Model> void syncPage(
+        GraphQLRequest<PaginatedResult<ModelWithMetadata<T>>> request,
+        SingleEmitter<Iterable<ModelWithMetadata<T>>> singleEmitter,
+        Set<ModelWithMetadata<T>> emittedValue) {
```

Comment:
I like to use +8 for method params, so that the body is clearly distinguished at +4
```suggestion
            GraphQLRequest<PaginatedResult<ModelWithMetadata<T>>> request,
            SingleEmitter<Iterable<ModelWithMetadata<T>>> singleEmitter,
            Set<ModelWithMetadata<T>> emittedValue) {
```

Reply:
Agreed on the indentation preference!  This method was refactored in my latest push though, so no longer relevant.